src/server/spatial_map.h

```diff
@@ -0,0 +1,83 @@
+/*
+Minetest
+Copyright (C) 2024, ExeVirus <nodecastmt@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#pragma once
+
+#include <functional>
+#include <vector>
+#include <unordered_set>
+#include "irrlichttypes_bloated.h"
+#include "constants.h"
+
+namespace server
+{
+class SpatialMap
+{
+public:
+	void insert(u16 id, const v3f &pos, bool postIteration = false);
+	void remove(u16 id, const v3f &pos, bool postIteration = false);
+	void remove(u16 id);
+	void removeAll();
+	void updatePosition(u16 id, const v3f &oldPos, const v3f &newPos);
+	void getRelevantObjectIds(const aabb3f &box, const std::function<void(u16 id)> &callback);
+	void handleInsertsAndDeletes();
+
+protected:
+	struct SpatialKey {
+		u16 padding_or_optional_id{0};
+		s16 x;
+		s16 y;
+		s16 z;
+
+		SpatialKey(s16 _x, s16 _y, s16 _z, bool _shrink = true) {
+			if(_shrink) {
+				x = _x >> 4;
+				y = _y >> 4;
+				z = _z >> 4;
+			} else {
+				x = _x;
+				y = _y;
+				z = _z;
+			}
+		}
+		SpatialKey(const v3f &_pos) : SpatialKey(_pos.X / BS, _pos.Y / BS, _pos.Z / BS){}
+		// The following use case is for storing pending insertions and deletions while iterating
+		// using the extra 16 bit padding makes keeping track of them super efficient for hashing.
+		SpatialKey(const v3f &_pos, const u16 id) : SpatialKey(_pos.X / BS, _pos.Y / BS, _pos.Z / BS, false){
+			padding_or_optional_id = id;
+		}
+
+		bool operator==(const SpatialKey &other) const {
+			return (x == other.x && y == other.y && z == other.z);
+		}
+	};
+
+	struct SpatialKeyHash {
+		auto operator()(const SpatialKey &key) const -> size_t {
+			return std::hash<size_t>()(*reinterpret_cast<const size_t*>(&key));
+		}
+	};
+
+	std::unordered_multimap<SpatialKey, u16, SpatialKeyHash> m_cached;
```

Comment:
Also, I think you could just use a map of vectors here, `std::unordered_map<SpatialKey, std::vector<u16>>`

I think it might be faster and have less memory overhead.

Reply:
This unordered multi-map is optimized for this kind of use case. I read through the GCC and VCC implementations (can't say I understand everything sadly). They have different optimizations for different sized buckets over plain old std::vector stores that allow it to be more performant. I believe their cutoffs were 16, and 64+ entries in a bucket for how they store things internally. 

It also has better allocator performance from what I could tell (both in RAM usage and latency), as I tested a unordered map of vectors myself with VCC. From what I could tell it came down to how they handle empty buckets/hashes, In the unordered_map of vectors, when I remove the last element from an existing vector in an unordered_map, It won't get de-allocated automatically, (more usage), and I still have to check it's size after I know it's empty. I was seeing that they do opimtizations in the unordered_multimap that made those cases much more performant and actually use less memory overhead.

Reply:
thank you for verifying that!